In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("bitext/bitext-gen-ai-chatbot-customer-support-dataset")

print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/


## Summary:

### Q&A
Yes, the dataset has been successfully loaded into the `df` DataFrame.

### Data Analysis Key Findings
*   The primary CSV file, `Bitext_Sample_Customer_Support_Training_Dataset_27K_responses-v11.csv`, was successfully identified and loaded into a pandas DataFrame named `df`.
*   The DataFrame `df` contains columns such as 'flags', 'instruction', 'category', 'intent', and 'response', as verified by displaying its head.

### Insights or Next Steps
*   The loaded DataFrame `df` is now ready for further data exploration, cleaning, and analysis to understand the customer support interactions.
*   A logical next step would be to perform a quick data overview, including checking data types, missing values, and unique values in categorical columns.


In [2]:
import pandas as pd
import os

# List files in the dataset directory
files_in_path = os.listdir(path)
print(f"Files available in the dataset directory: {files_in_path}")

# Identify the main CSV file. Based on previous runs or common dataset structures,
# it's often 'Bitext_Sample_-_es_to_en_-_Sheet1.csv'
# Let's try to find a .csv file or assume the given name.
csv_file_name = None
for file in files_in_path:
    if file.endswith('.csv'):
        csv_file_name = file
        break

if csv_file_name is None:
    print("No CSV file found in the dataset directory. Attempting to use a known filename.")
    # Fallback to a known filename if no CSV is automatically detected
    csv_file_name = 'Bitext_Sample_-_es_to_en_-_Sheet1.csv'

csv_file_path = os.path.join(path, csv_file_name)

# Load the CSV file into a pandas DataFrame
df = pd.read_csv(csv_file_path)

print(f"Successfully loaded '{csv_file_name}' into DataFrame 'df'.")

# Display the head of the DataFrame to verify
df.head()

Files available in the dataset directory: ['Bitext_Sample_Customer_Support_Training_Dataset_27K_responses-v11.csv']
Successfully loaded 'Bitext_Sample_Customer_Support_Training_Dataset_27K_responses-v11.csv' into DataFrame 'df'.


flags                                        instruction category  \
0      B   question about cancelling order {{Order Number}}    ORDER   
1    BQZ  i have a question about cancelling oorder {{Or...    ORDER   
2   BLQZ    i need help cancelling puchase {{Order Number}}    ORDER   
3     BL         I need to cancel purchase {{Order Number}}    ORDER   
4  BCELN  I cannot afford this order, cancel purchase {{...    ORDER   

         intent                                           response  
0  cancel_order  I've understood you have a question regarding ...  
1  cancel_order  I've been informed that you have a question ab...  
2  cancel_order  I can sense that you're seeking assistance wit...  
3  cancel_order  I understood that you need assistance with can...  
4  cancel_order  I'm sensitive to the fact that you're facing f...

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26872 entries, 0 to 26871
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   flags        26872 non-null  object
 1   instruction  26872 non-null  object
 2   category     26872 non-null  object
 3   intent       26872 non-null  object
 4   response     26872 non-null  object
dtypes: object(5)
memory usage: 1.0+ MB


In [4]:
import re
import random



# Define the regex pattern for placeholders
placeholder_pattern = r"\{\{([^}]+)\}\}"

# Initialize a set to store unique placeholders
unique_placeholders = set()

# Iterate through the 'instruction' column and extract placeholders
if 'instruction' in df.columns:
    for instruction_text in df['instruction'].astype(str):
        matches = re.findall(placeholder_pattern, instruction_text)
        for match in matches:
            unique_placeholders.add(match.strip())

    print("Unique placeholders found:")
    for placeholder in sorted(list(unique_placeholders)):
        print(placeholder)
else:
    print("The 'instruction' column does not exist in the DataFrame.")

Unique placeholders found:
Account Category
Account Type
Currency Symbol
Delivery City
Delivery Country
Invoice Number
Order Number
Person Name
Refund Amount


In [5]:
import re

# Define the mapping of placeholders to synthetic values
synthetic_values = {
    "Order Number": "ORD-2025-12345",
    "Person Name": "Maria Garcia",
    "email": "alex.smith@example.com", # Assuming 'email' is a placeholder
    "product_name": "Wireless Mouse", # Assuming 'product_name' is a placeholder
    "date": "05/09/2025", # Assuming 'date' is a placeholder
    "Refund Amount": "$50.00", # Based on previously found placeholders
    "Account Category": "Savings", # Based on previously found placeholders
    "Account Type": "Checking", # Based on previously found placeholders
    "Currency Symbol": "$", # Based on previously found placeholders
    "Delivery City": "New York", # Based on previously found placeholders
    "Delivery Country": "USA", # Based on previously found placeholders
    "Invoice Number": "INV-2025-54321", # Based on previously found placeholders
}

# Create a copy of the DataFrame to store the modified instructions
df_synthetic = df.copy()

# Function to replace placeholders
def replace_placeholders(text):
    for placeholder, value in synthetic_values.items():
        # Use re.escape to handle special characters in placeholder names
        text = re.sub(r'\{\{' + re.escape(placeholder) + r'\}\}', value, text, flags=re.IGNORECASE)
    return text

# Apply the replacement function to the 'instruction' column
if 'instruction' in df_synthetic.columns:
    df_synthetic['instruction_synthetic'] = df_synthetic['instruction'].apply(replace_placeholders)
    print("Instructions with synthetic values generated successfully. Displaying original and synthetic instructions for verification:")
    display(df_synthetic[['instruction', 'instruction_synthetic']].head())
else:
    print("The 'instruction' column does not exist in the DataFrame.")

Instructions with synthetic values generated successfully. Displaying original and synthetic instructions for verification:


instruction  \
0   question about cancelling order {{Order Number}}   
1  i have a question about cancelling oorder {{Or...   
2    i need help cancelling puchase {{Order Number}}   
3         I need to cancel purchase {{Order Number}}   
4  I cannot afford this order, cancel purchase {{...   

                               instruction_synthetic  
0     question about cancelling order ORD-2025-12345  
1  i have a question about cancelling oorder ORD-...  
2      i need help cancelling puchase ORD-2025-12345  
3           I need to cancel purchase ORD-2025-12345  
4  I cannot afford this order, cancel purchase OR...

In [6]:
df_synthetic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26872 entries, 0 to 26871
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   flags                  26872 non-null  object
 1   instruction            26872 non-null  object
 2   category               26872 non-null  object
 3   intent                 26872 non-null  object
 4   response               26872 non-null  object
 5   instruction_synthetic  26872 non-null  object
dtypes: object(6)
memory usage: 1.2+ MB


In [7]:
if 'instruction_synthetic' in df_synthetic.columns:
    print("Sampling 20 processed rows for verification:")
    display(df_synthetic[['instruction', 'instruction_synthetic']].sample(n=20, random_state=42))
else:
    print("The 'instruction_synthetic' column does not exist in the DataFrame. Please ensure the placeholder replacement step was executed correctly.")

Sampling 20 processed rows for verification:


instruction  \
9329                    I can't talk with  a human agent   
4160   I have got to locate hte bills from {{Person N...   
18500  I cannot pay, help me to inform of a problem w...   
8840            I want help speaking to customer service   
5098            I try to see th accepted payment options   
17250        where to sign up to the company nmewsletter   
3589   I'd like to see the withdrwaal fee how can i d...   
9043                        I want to speak with someone   
15800               can you help me getting bill #85632?   
4384   I don't know how to take a quick look at invoi...   
11150     I don't know how to delete my platinum account   
6417   help me check in what cases can I ask to be re...   
4186   is it possible to locate my bill from {{Person...   
7301   i want help to file a consumer claim against u...   
8267        uhave a free number to call customer support   
3798   i do not know how i can see the early terminat...   
17355  could you help me to unsubscribe to the damn n...   
21099  I don't know how to inform of problems with a ...   
15331                 download bill from {{Person Name}}   
15938        want assistance receiving a refund of money   

                                   instruction_synthetic  
9329                    I can't talk with  a human agent  
4160    I have got to locate hte bills from Maria Garcia  
18500  I cannot pay, help me to inform of a problem w...  
8840            I want help speaking to customer service  
5098            I try to see th accepted payment options  
17250        where to sign up to the company nmewsletter  
3589   I'd like to see the withdrwaal fee how can i d...  
9043                        I want to speak with someone  
15800               can you help me getting bill #85632?  
4384   I don't know how to take a quick look at invoi...  
11150     I don't know how to delete my platinum account  
6417   help me check in what cases can I ask to be re...  
4186   is it possible to locate my bill from Maria Ga...  
7301   i want help to file a consumer claim against u...  
8267        uhave a free number to call customer support  
3798   i do not know how i can see the early terminat...  
17355  could you help me to unsubscribe to the damn n...  
21099  I don't know how to inform of problems with a ...  
15331                    download bill from Maria Garcia  
15938        want assistance receiving a refund of money

In [8]:
from sklearn.model_selection import train_test_split

# Drop the original 'instruction' column and keep 'instruction_synthetic'
if 'instruction' in df_synthetic.columns:
    df_synthetic = df_synthetic.drop(columns=['instruction'])

# Define the features (X) and target (y) for intent classification
X = df_synthetic.drop(columns=['response', 'intent']) # Features should not include the target or the response
y = df_synthetic['intent'] # Target is the 'intent' column

# First split: 80% train, 20% (validation + test)
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Second split: 50% validation, 50% test from the 20% temporary set
# This results in 10% validation and 10% test from the original dataset
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42, stratify=y_temp)

print(f"Training set size: {len(X_train)} samples")
print(f"Validation set size: {len(X_val)} samples")
print(f"Test set size: {len(X_test)} samples")

# Display the head of the training set features as an example
print("\nHead of X_train:")
display(X_train.head())

Training set size: 21497 samples
Validation set size: 2687 samples
Test set size: 2688 samples

Head of X_train:


flags  category                              instruction_synthetic
18407    BL   PAYMENT  I want assistance informing of a trouble with ...
4878    BKL   INVOICE                                    see bill #12588
26858   BCL    REFUND    I want to see my refund current status, help me
25464   BKL     ORDER         see current status of order ORD-2025-12345
12237  BCIL  DELIVERY  I need to check the options for shipping, how ...

Projects Goal is to -
Build, evaluate, and compare two modeling pathways for intent classification —
LLM-based zero/few-shot inference and fine-tuned transformers; both using uniform
evaluation metrics (Accuracy, Macro-F1, and Cost).

Approach 1: Zero- or Few-Shot via Large Language Models

Sample Test Data for LLM Experiments: Sample up to 10 examples from each unique intent class from the X_test and y_test datasets, storing them in X_test_sampled and y_test_sampled respectively. This ensures a controlled and representative dataset for LLM experiments while managing API costs.

Define Zero-Shot Prompt Structure: Create a Python function to construct the zero-shot prompt for the LLM. This prompt will use the combined_instruction from the sampled data and explicitly instruct the LLM to classify the intent, potentially listing the available intent classes to guide its response.

Implement LLM Inference Function: Provide a placeholder or example function that demonstrates how to make an LLM API call (e.g., using OpenAI's API client). This function will take a prompt and return the LLM's response. It will also include guidance for the user to insert their actual API key and handle the specific API integration for their chosen LLM, emphasizing careful usage to manage costs.

Run LLM Inference and Collect Predictions: Iterate through the sampled test data, generate a zero-shot prompt for each entry, call the LLM inference function to get a prediction, and collect all predicted intents. This step will also include parsing the LLM's response to extract the intent label.

Evaluate LLM Performance: Calculate and display the Accuracy and Macro-F1 scores by comparing the collected LLM predictions against the true intent labels from the sampled test data. This will provide the initial performance metrics for the zero-shot approach.

Final Task: Summarize the results of the zero-shot LLM intent classification and prepare for the next steps, potentially exploring few-shot examples or moving to the fine-tuned transformer approach.



# Task
Prepare the LLM environment, sample 10 examples from each unique intent class from `X_test` and `y_test` for LLM experiments, and then implement a zero-shot intent classification workflow. This workflow should include defining a prompt structure using `combined_instruction`, developing a function for LLM inference, running the inference to collect predictions, and finally evaluating the LLM's performance using Accuracy and Macro-F1 scores.

## Prepare LLM Environment and Sample Data

### Subtask:
Guide the user to set up their chosen LLM API key (e.g., as an environment variable) and then sample 10 examples from each unique intent class from the `X_test` and `y_test` datasets. This sampled data will be used for all LLM experiments to control costs.


**Reasoning**:
Now that the instructions for setting up the API key are provided, the next step is to implement the sampling logic. This involves identifying unique intents, sampling up to 10 examples per intent from the test sets, and combining them into new sampled DataFrames for features and labels.



# Task
```python
import pandas as pd

# Ensure 'intent' column is available in X_test for sampling
if 'intent' not in X_test.columns:
    print("Error: 'intent' column not found in X_test. Please ensure it's present for stratified sampling.")
else:
    # Get unique intent classes
    unique_intents = X_test['intent'].unique()
    print(f"Found {len(unique_intents)} unique intent classes.")

    X_test_sampled_list = []
    y_test_sampled_list = []

    # Iterate through each unique intent to sample examples
    for intent_class in unique_intents:
        # Filter X_test and y_test for the current intent class
        X_intent = X_test[X_test['intent'] == intent_class]
        y_intent = y_test[X_test['intent'] == intent_class] # y_test should align with X_test indices

        # Sample up to 10 examples from the current intent class
        # Use min(len(X_intent), 10) to handle classes with fewer than 10 examples
        if not X_intent.empty:
            sampled_indices = X_intent.sample(n=min(len(X_intent), 10), random_state=42).index
            X_test_sampled_list.append(X_test.loc[sampled_indices])
            y_test_sampled_list.append(y_test.loc[sampled_indices])

    # Concatenate all sampled data into new DataFrames
    X_test_sampled = pd.concat(X_test_sampled_list)
    y_test_sampled = pd.concat(y_test_sampled_list)

    print(f"\nTotal sampled examples: {len(X_test_sampled)}")
    print(f"Number of unique intents in sampled data: {X_test_sampled['intent'].nunique()}")

    print("\nHead of X_test_sampled:")
    display(X_test_sampled.head())

    print("\nValue counts for 'intent' in X_test_sampled to verify distribution:")
    display(X_test_sampled['intent'].value_counts())

    print("\nHead of y_test_sampled:")
    display(y_test_sampled.head())
```

Sample Test Data for LLM Experiments: Sample up to 10 examples from each unique intent class from the X_test and y_test datasets, storing them in X_test_sampled and y_test_sampled respectively. This ensures a controlled and representative dataset for LLM experiments while managing API costs.

We removed 'intent' and 'Respone' from X, should category be also removed or would it be data leakage? What is Data Leakage? Data leakage occurs when information from outside the training dataset is used to create the model, leading to overly optimistic performance estimates. In classification tasks, this often happens when a feature directly or indirectly reveals the target variable.

The 'category' Column in our X Features: Currently, the X features (which are used to construct the combined_instruction for the LLM prompt) include flags, category, and instruction_synthetic. So, yes, the category column is being fed to the LLM as part of the input prompt.

Does 'category' cause leakage here? It could lead to a form of leakage if the 'category' is a direct, one-to-one mapping with the 'intent'. For example, if there's only one intent per category, then knowing the category directly tells you the intent.

However, in many real-world scenarios (and seemingly in this dataset based on the examples), 'category' provides broader context (e.g., 'ORDER' category can contain intents like cancel_order, track_order, change_order). In such cases, category acts as a helpful contextual feature that allows the LLM to narrow down the possibilities without explicitly revealing the exact intent. When building a prompt for an LLM, providing relevant contextual information like the category is often beneficial for its reasoning.

For the purpose of evaluating the zero-shot LLM's ability to classify intent given this context, including 'category' in the prompt is a reasonable approach to provide the LLM with sufficient information to make its prediction. If we were building a traditional machine learning model from scratch and trying to predict intent purely from the instruction_synthetic without any pre-categorization, we would need to carefully analyze the correlation between 'category' and 'intent' to decide if 'category' should be an independent feature or if it leads to excessive leakage.

For now, we'll continue using category as a contextual element in the LLM prompt, as it is a common practice to provide LLMs with as much relevant input as possible to guide their zero-shot reasoning.



In [9]:
import pandas as pd

# The 'intent' column is the target variable y_test, not in X_test.
# We need to sample based on y_test and then use the sampled indices to get corresponding X_test rows.

# Get unique intent classes from y_test
unique_intents = y_test.unique()
print(f"Found {len(unique_intents)} unique intent classes.")

X_test_sampled_list = []
y_test_sampled_list = []

# Iterate through each unique intent to sample examples
for intent_class in unique_intents:
    # Filter y_test for the current intent class to get its indices
    intent_indices = y_test[y_test == intent_class].index

    # Sample up to 10 indices from the current intent class
    # Use min(len(intent_indices), 10) to handle classes with fewer than 10 examples
    if not intent_indices.empty:
        # Sample indices directly from the filtered indices
        sampled_indices = intent_indices.to_series().sample(n=min(len(intent_indices), 10), random_state=42).index
        X_test_sampled_list.append(X_test.loc[sampled_indices])
        y_test_sampled_list.append(y_test.loc[sampled_indices])

# Concatenate all sampled data into new DataFrames
X_test_sampled = pd.concat(X_test_sampled_list)
y_test_sampled = pd.concat(y_test_sampled_list)

print(f"\nTotal sampled examples: {len(X_test_sampled)}")
print(f"Number of unique intents in sampled data: {y_test_sampled.nunique()}") # Check unique intents in y_test_sampled

print("\nHead of X_test_sampled:")
display(X_test_sampled.head())

print("\nValue counts for 'intent' in y_test_sampled to verify distribution:")
display(y_test_sampled.value_counts()) # Display value counts from y_test_sampled

print("\nHead of y_test_sampled:")
display(y_test_sampled.head())

Found 27 unique intent classes.

Total sampled examples: 270
Number of unique intents in sampled data: 27

Head of X_test_sampled:


flags category                              instruction_synthetic
1267  BLQZ    ORDER  i have got to swap  an item of order ORD-2025-...
1179   BIL    ORDER          how do I correct purchase ORD-2025-12345?
1310  BILZ    ORDER  canm you help me to update purchase ORD-2025-1...
1048  BLMQ    ORDER  i try to switch several items of order ORD-202...
1950    BL    ORDER  I need information about correcting purchase O...


Value counts for 'intent' in y_test_sampled to verify distribution:


intent
change_order                10
newsletter_subscription     10
review                      10
track_refund                10
delivery_options            10
create_account              10
delivery_period             10
contact_human_agent         10
payment_issue               10
complaint                   10
check_cancellation_fee      10
set_up_shipping_address     10
track_order                 10
cancel_order                10
edit_account                10
recover_password            10
get_invoice                 10
switch_account              10
registration_problems       10
contact_customer_service    10
place_order                 10
change_shipping_address     10
check_invoice               10
get_refund                  10
check_refund_policy         10
check_payment_methods       10
delete_account              10
Name: count, dtype: int64


Head of y_test_sampled:


1267    change_order
1179    change_order
1310    change_order
1048    change_order
1950    change_order
Name: intent, dtype: object

Unique Intents Found: We identified all 27 unique intent classes from your y_test dataset.
Total Sampled Examples: A total of 270 examples have been sampled.
Stratified Sampling: As intended, the value_counts for y_test_sampled confirms that we have exactly 10 samples for each of the 27 unique intent classes, ensuring a balanced representation for LLM evaluation.
This means our sampled dataset is now perfectly prepared for the next step: defining the zero-shot prompt structure for the LLM!


Let's redefine the zero-shot prompt structure. This updated prompt will explicitly instruct the LLM to identify the intent from the customer query and provide the list of possible intent categories it should choose from. After the modification, I will show an example of the generated prompt.

In [77]:
import pandas as pd

# Get the list of all unique intent classes from the full dataset
all_unique_intents = df_synthetic['intent'].unique()
all_unique_intents_sorted = sorted(all_unique_intents)
# Format as a bulleted list for better LLM parsing
all_unique_intents_bulleted = "\n- " + "\n- ".join(all_unique_intents_sorted)

print(f"All unique intents available (formatted for prompt): {all_unique_intents_bulleted}\n")

def create_zero_shot_prompt(category: str, instruction_synthetic: str, available_intents_bulleted: str) -> str:
    """
    Constructs a zero-shot prompt for intent classification with controlled response format.
    """

    prompt = f'''You are an intent classification system.
Your task is to classify the intent of a customer query.

Specific guidelines (but not rules)

Customers expecting a refund is not inherently getting a refund. They may be expecting one.
Customers asking to check when their order will arrive are asking about the delivery period - unless an order number is apart of the instruction.
Customers asking about possibility of orders are asking about delivery options, including where they are ordering from.
Customers asking setting up a new or secondary address are not editing their address, but setting up an address.
Customers looking to use a specific account (such as chequing and savings) or profile, are attempting to switch accounts.
Customers looking customer assistance aren't inherently looking for human assistance.
If an invoice has already been issued, and the customer is looking for it, that is check_invoice. 
Refund policy questions relate to check_refund_policy. Note that if they aren't asking for specifics on the policy, they are likely just wanting to get a refund.


Customer Query Details:
- Category: {category}
- User Instruction: {instruction_synthetic}

You MUST choose exactly one intent ONLY from the following list:

Available Intents:
{available_intents_bulleted}

### Response Format (follow EXACTLY):
1. Provide an explanation (1-2 paragraphs) of why you selected the intent. Be sure to consider two-to-three other intents before settling on a final choice.
2. Then output the final predicted intent on a new line in the following format:

Final Choice: <intent_label>

Do NOT invent new intents.
Do NOT output anything other than the two required parts.

Begin.

Explanation:'''
    return prompt


# Example usage with a sample from X_test_sampled
if not X_test_sampled.empty:
    sample_category = X_test_sampled['category'].iloc[0]
    sample_instruction_synthetic = X_test_sampled['instruction_synthetic'].iloc[0]
    example_prompt = create_zero_shot_prompt(sample_category, sample_instruction_synthetic, all_unique_intents_bulleted)
    print("Example Zero-Shot Prompt (improved):\n")
    print(example_prompt)
else:
    print("X_test_sampled is empty. Cannot generate an example prompt.")

All unique intents available (formatted for prompt): 
- cancel_order
- change_order
- change_shipping_address
- check_cancellation_fee
- check_invoice
- check_payment_methods
- check_refund_policy
- complaint
- contact_customer_service
- contact_human_agent
- create_account
- delete_account
- delivery_options
- delivery_period
- edit_account
- get_invoice
- get_refund
- newsletter_subscription
- payment_issue
- place_order
- recover_password
- registration_problems
- review
- set_up_shipping_address
- switch_account
- track_order
- track_refund

Example Zero-Shot Prompt (improved):

You are an intent classification system.
Your task is to classify the intent of a customer query.

Specific guidelines (but not rules)

Customers expecting a refund is not inherently getting a refund. They may be expecting one.
Customers asking to check when their order will arrive are asking about the delivery period - unless an order number is apart of the instruction.
Customers asking about possibility o

## Implement LLM Inference Function

### Subtask:
Provide a placeholder or example function that demonstrates how to make an LLM API call. This function will take a prompt and return the LLM's response. It will also include guidance for the user to insert their actual API key and handle the specific API integration for their chosen LLM, emphasizing careful usage to manage costs.

In [78]:
import os
import time
import re
from openai import OpenAI, APIError, RateLimitError
from kaggle_secrets import UserSecretsClient

user_secrets = UserSecretsClient()
OPENAI_API_KEY = user_secrets.get_secret("OPENAI_API_KEY")

client = OpenAI(api_key=OPENAI_API_KEY)

def safe_llm_predict(
    prompt: str,
    model_name: str = "gpt-4.1-mini",
    timeout: int = 60,
    max_retries: int = 6
):

    input_tokens = 0
    output_tokens = 0
    inference_time = 0.0
    cost = 0.0
    extracted_intent = "LLM Error"

    for attempt in range(max_retries):
        start_time = time.time()

        try:
            response = client.chat.completions.create(
                model=model_name,
                messages=[{"role": "user", "content": prompt}],
                timeout=timeout
            )

            inference_time = time.time() - start_time

            # FIXED: message is an object, not a dict
            raw_text = response.choices[0].message.content.strip()

            # usage handling
            if response.usage:
                input_tokens = response.usage.prompt_tokens or 0
                output_tokens = response.usage.completion_tokens or 0


            # Strong regex match
            match = re.search(r"Final\s*Choice\s*:\s*([A-Za-z0-9_\-]+)",
                              raw_text, flags=re.IGNORECASE)
            if match:
                extracted_intent = match.group(1).strip()
                return extracted_intent, inference_time, input_tokens, output_tokens, cost

            # Fallback: take final word after "Final Choice:"
            match = re.search(r"Final\s*Choice\s*:\s*(.*)$",
                              raw_text, flags=re.IGNORECASE | re.DOTALL)
            if match:
                tail = match.group(1).strip()
                last_word = re.findall(r"[A-Za-z0-9_\-]+", tail)
                if last_word:
                    extracted_intent = last_word[-1]
                    return extracted_intent, inference_time, input_tokens, output_tokens, cost

            # Full fallback
            fallback_tokens = re.findall(r"[A-Za-z0-9_\-]+", raw_text)
            if fallback_tokens:
                extracted_intent = fallback_tokens[-1]

            return extracted_intent, inference_time, input_tokens, output_tokens, cost

        except RateLimitError:
            wait_time = 1.5 ** attempt
            print(f"[429] Rate limit — retrying in {wait_time:.2f}s...")
            time.sleep(wait_time)
            continue

        except APIError as e:
            return f"LLM Error: {e}", inference_time, input_tokens, output_tokens, cost

        except Exception as e:
            return f"LLM Error: {e}", inference_time, input_tokens, output_tokens, cost

    return extracted_intent, inference_time, input_tokens, output_tokens, cost


Here's a breakdown of the output:

"Starting LLM inference for 270 samples... LLM inference completed.": This confirms that the loop processed all 270 sampled test cases, generating a prediction for each one using the gemini-2.0-flash-lite model.

"Head of LLM Predictions:"

This displays the first few rows of the predictions_df DataFrame. You can see the original index of the sampled X_test examples and the corresponding llm_predicted_intent. For instance, you can see several change_order predictions at the top.
"Value counts of LLM Predictions:"

This is a crucial summary, showing the frequency of each intent predicted by the LLM across all 270 samples. Ideally, if the LLM was perfect and predicted correctly for all 10 samples of each of the 27 intents, we would see each unique intent listed with a count of 10. This output helps us quickly identify which intents the LLM is predicting more often, or if there are any unexpected predictions or errors.
This output now gives us the LLM's zero-shot predictions, which we can compare against the true y_test_sampled labels to evaluate its performance!



In [79]:
def test_safe_llm_predict(prompt: str):
    """
    Direct test harness to inspect what the model returns
    before running the full dataset loop.
    """

    print("\n======================")
    print("🔍 TESTING LLM OUTPUT")
    print("======================\n")

    # Make a single request
    intent, inf_time, in_tokens, out_tokens, cost = safe_llm_predict(prompt)

    print("📌 Extracted Intent:", intent)
    print("⏱️ Inference Time:", inf_time)
    print("🔢 Input Tokens:", in_tokens)
    print("🔢 Output Tokens:", out_tokens)
    print("💲 Cost:", cost)

    print("\n--- RAW MODEL OUTPUT ---\n")

    # Call the API *again* but without extraction
    # so you can see EXACTLY what the model said
    raw = client.chat.completions.create(
        model="gpt-4.1-mini",
        messages=[{"role": "user", "content": prompt}]
    )

    raw_text = raw.choices[0].message.content
    print(raw_text)
    print("\n------------------------\n")

    return raw_text


In [80]:
# Pick a sample row to test
sample_row = X_test_sampled.iloc[0]

test_prompt = create_zero_shot_prompt(
    sample_row["category"],
    sample_row["instruction_synthetic"],
    all_unique_intents_bulleted
)

# Run the single test
test_safe_llm_predict(test_prompt)



🔍 TESTING LLM OUTPUT

📌 Extracted Intent: change_order
⏱️ Inference Time: 3.017700672149658
🔢 Input Tokens: 460
🔢 Output Tokens: 183
💲 Cost: 0.0

--- RAW MODEL OUTPUT ---

Explanation:  
The customer query explicitly states they want to "swap an item" from an existing order, which implies they want to modify or change something already ordered. This aligns closely with the intent "change_order," as it involves adjusting the content of the order rather than canceling it altogether or merely tracking it. The query does mention a specific order number but does not request information about delivery periods or tracking, so "delivery_period" and "track_order" are less appropriate. Although the query involves an exchange, it does not mention refunds directly, so "get_refund" is unlikely. Therefore, the best fit is "change_order" as the customer wants to alter the items within an order.

Final Choice: change_order

------------------------



'Explanation:  \nThe customer query explicitly states they want to "swap an item" from an existing order, which implies they want to modify or change something already ordered. This aligns closely with the intent "change_order," as it involves adjusting the content of the order rather than canceling it altogether or merely tracking it. The query does mention a specific order number but does not request information about delivery periods or tracking, so "delivery_period" and "track_order" are less appropriate. Although the query involves an exchange, it does not mention refunds directly, so "get_refund" is unlikely. Therefore, the best fit is "change_order" as the customer wants to alter the items within an order.\n\nFinal Choice: change_order'

In [81]:
from tqdm import tqdm
import time
import pandas as pd

# Storage for metrics
llm_predictions = []
inference_times = []
input_token_counts = []
output_token_counts = []
api_costs = []

print(f"Starting LLM inference for {len(X_test_sampled)} samples...\n")

# tqdm progress bar
for idx, row in tqdm(X_test_sampled.iterrows(), total=len(X_test_sampled), desc="LLM Inference"):
    category = row["category"]
    instruction_synthetic = row["instruction_synthetic"]

    # Build zero-shot prompt
    prompt = create_zero_shot_prompt(
        category,
        instruction_synthetic,
        all_unique_intents_bulleted
    )

    # Call LLM safely
    prediction, inf_time, in_tokens, out_tokens, call_cost = safe_llm_predict(
        prompt,
        timeout=60
    )

    # Store metrics
    llm_predictions.append(prediction)
    inference_times.append(inf_time)
    input_token_counts.append(in_tokens)
    output_token_counts.append(out_tokens)
    api_costs.append(call_cost)

    # Slight delay to avoid rate limits
    time.sleep(0.35)

print("\nLLM inference completed.\n")

# Convert predictions to DataFrame
predictions_df = pd.DataFrame(
    {"llm_predicted_intent": llm_predictions},
    index=X_test_sampled.index
)

print("Head of LLM Predictions:")
display(predictions_df.head())

print("\nValue counts of LLM Predictions:")
display(predictions_df["llm_predicted_intent"].value_counts())


Starting LLM inference for 270 samples...



LLM Inference: 100%|██████████| 270/270 [12:39<00:00,  2.81s/it]


LLM inference completed.

Head of LLM Predictions:


llm_predicted_intent
1267         change_order
1179         change_order
1310         change_order
1048         change_order
1950         change_order


Value counts of LLM Predictions:


llm_predicted_intent
get_refund                  13
contact_customer_service    11
delivery_options            11
check_invoice               11
contact_human_agent         10
change_order                10
newsletter_subscription     10
payment_issue               10
delivery_period             10
complaint                   10
create_account              10
delete_account              10
switch_account              10
set_up_shipping_address     10
check_cancellation_fee      10
cancel_order                10
edit_account                10
recover_password            10
track_order                 10
check_payment_methods       10
change_shipping_address     10
registration_problems       10
review                       9
place_order                  9
get_invoice                  9
check_refund_policy          9
track_refund                 8
Name: count, dtype: int64

In [85]:
# Build a dataframe with all relevant info
results_df = pd.DataFrame({
    "y_true": y_true,
    "y_pred": y_pred,
    "instruction_synthetic": X_test_sampled["instruction_synthetic"].values,
    "category": X_test_sampled["category"].values
})

# Filter wrong predictions
wrong_df = results_df[results_df["y_true"] != results_df["y_pred"]]

print("Number of wrong predictions:", len(wrong_df))
wrong_df.head(27)

Number of wrong predictions: 7


y_true                    y_pred  \
22042               review  contact_customer_service   
26158         track_refund                get_refund   
26327         track_refund                get_refund   
26265         track_refund                get_refund   
15126          get_invoice             check_invoice   
19473          place_order          delivery_options   
6120   check_refund_policy              track_refund   

                                   instruction_synthetic  category  
22042  need assistance leaving some feedback about yo...  FEEDBACK  
26158           i expect a restitution of $50.00 dollars    REFUND  
26327                       i expect a rebate of $$50.00    REFUND  
26265          i expect a compensqtion of $50.00 dollars    REFUND  
15126  i have to download the invoices from Maria Garcia   INVOICE  
19473    is it possible to acquire several of ur article     ORDER  
6120   can I check how long reimbursements usually take?    REFUND

In [84]:
import pandas as pd
from sklearn.metrics import accuracy_score, f1_score

# Ensure y_test_sampled and predictions_df are aligned by index
y_true = y_test_sampled.loc[predictions_df.index]
y_pred = predictions_df['llm_predicted_intent']

# Calculate Accuracy
accuracy = accuracy_score(y_true, y_pred)

# Calculate Macro-F1 Score
f1_macro = f1_score(y_true, y_pred, average='macro')

print(f"Zero-Shot LLM Performance:\n")
print(f"Accuracy: {accuracy:.4f}")
print(f"Macro-F1 Score: {f1_macro:.4f}")

# Calculate and print additional metrics
total_inference_time = sum(inference_times)
average_time_per_sample = total_inference_time / len(X_test_sampled)
estimated_time_per_1000_samples = average_time_per_sample * 1000

total_input_tokens = sum(input_token_counts)
total_output_tokens = sum(output_token_counts)
total_api_cost = sum(api_costs)

print(f"\n--- LLM Inference Metrics ---")
print(f"Total samples processed: {len(X_test_sampled)}")
print(f"Total inference time: {total_inference_time:.2f} seconds")
print(f"Average inference time per sample: {average_time_per_sample:.4f} seconds")
print(f"Estimated inference time per 1000 samples: {estimated_time_per_1000_samples:.2f} seconds")
print(f"Total input tokens consumed: {total_input_tokens}")
print(f"Total output tokens consumed: {total_output_tokens}")
print(f"Approximate API Cost for {len(X_test_sampled)} samples: ${total_api_cost:.6f}")
print(f"Estimated API Cost per 1000 samples: ${total_api_cost / len(X_test_sampled) * 1000:.6f}")

Zero-Shot LLM Performance:

Accuracy: 0.9741
Macro-F1 Score: 0.9739

--- LLM Inference Metrics ---
Total samples processed: 270
Total inference time: 664.41 seconds
Average inference time per sample: 2.4608 seconds
Estimated inference time per 1000 samples: 2460.79 seconds
Total input tokens consumed: 122311
Total output tokens consumed: 40502
Approximate API Cost for 270 samples: $0.000000
Estimated API Cost per 1000 samples: $0.000000


In [56]:
X_test_sampled.head()

flags category                              instruction_synthetic
1267  BLQZ    ORDER  i have got to swap  an item of order ORD-2025-...
1179   BIL    ORDER          how do I correct purchase ORD-2025-12345?
1310  BILZ    ORDER  canm you help me to update purchase ORD-2025-1...
1048  BLMQ    ORDER  i try to switch several items of order ORD-202...
1950    BL    ORDER  I need information about correcting purchase O...